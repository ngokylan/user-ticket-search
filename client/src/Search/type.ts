import { FocusEventHandler, KeyboardEventHandler } from 'react';

export type SearchProps = {
  /** If search component is visible */
  isVisible: boolean;
  /** On close event handler */
  onClose?: Function;
  /** On search text box change event handler */
  onChange?: (newValue: string) => void;
  /** On search text box key press event handler */
  onKeyPress?: KeyboardEventHandler;
  /** Search text box value */
  value?: null | string;
  /** id attribute */
  id?: string;
  /** class name attribute */
  className?: string;
  /** Search text box on blur event handler */
  onBlur?: FocusEventHandler;
  /** Search text box on focus event handler */
  onFocus?: FocusEventHandler;
  /** Search text box on submit event handler */
  onSubmit?: Function;
  /** Define the placeholder of search */
  placeholder?: string;
  /** Add ariaLabel for WCAG accessibility compliance, (optional) */
  ariaLabel?: string;
  /** To make the submit button enabled always */
  isSubmitEnabledAlways?: boolean;
};
