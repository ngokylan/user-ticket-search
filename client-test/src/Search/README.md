
Search

```js
<Search
  isVisible={true}
  placeholder={'Search Placeholder'}
/>
```

```js
<Search
  value={'Search Keyword'}
  isVisible={true}
  isSubmitVisibleAlways={true}
/>
```