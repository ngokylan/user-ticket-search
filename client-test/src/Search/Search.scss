@import '../_lib/style/index';
$search: '.search';

#{$search} {
  order: 0;
  width: 100%;
  transition: all 0.25s;
  overflow: hidden;
  transform: translateY(-100%);

  &__wrapper {
    width: inherit;
    box-shadow: none;
    background-color: #fff;
    top: 0;
    padding: 8px 0 8px 12px;

    @include mobile {
      padding: 8px 32px;
    }

    @include tablet {
      padding: 16px 40px;
    }

    @include desktop {
      padding: 24px 40px;
    }
  }

  &__input {
    position: relative;

    .input {
      height: 48px;
      padding-right: 96px;
    }

    button.btn {
      position: absolute;
      right: 1px;
      top: 1px;
      height: 46px;
      border: none;
      background-color: #fff;

      &.search__btn-clear {
        right: 49px;
      }
    }
  }
}

// Modifier
#{$search}--open {
  transform: translateY(0);
}
