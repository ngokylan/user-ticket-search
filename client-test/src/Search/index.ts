export { default } from './Search';
export * from './type';
