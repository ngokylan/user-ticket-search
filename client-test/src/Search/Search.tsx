import React, { ChangeEvent, Component } from 'react';
import SearchClearButton from './component/SearchClearButton';
import SearchCloseButton from './component/SearchCloseButton';
import SearchSubmitButton from './component/SearchSubmitButton';
import { getClass } from '../_lib/helper';
import { SearchProps } from './type';
import Input from '../Input';
import Row from '../Grid/Row';
import Col from '../Grid/Col';
import './Search.scss';

class Search extends Component<SearchProps> {
  static displayName = 'Search';
  static defaultProps = {
    placeholder: 'Search',
    ariaLabel: 'Search field',
  };

  handleOnClose = () => {
    const { onClose } = this.props;
    if (onClose) {
      onClose();
    }
  };

  handleOnSubmit = () => {
    const { onSubmit } = this.props;
    if (onSubmit) {
      onSubmit();
    }
  };

  handleOnChange = (e: ChangeEvent<HTMLInputElement>) => {
    this.setValue(e.target.value);
  };

  handleOnClear = () => {
    this.setValue('');
  };

  setValue(value: string) {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
  }

  get value() {
    return this.props.value || '';
  }

  render() {
    const {
      onKeyPress,
      id,
      onFocus,
      onBlur,
      isVisible,
      onClose,
      placeholder,
      ariaLabel,
      isSubmitEnabledAlways,
    } = this.props;
    const { value } = this;

    return (
      <div
        id={id}
        className={getClass('search', {}, { open: isVisible })}
        data-testid={`search-${id || 'default'}`}
      >
        <div className="search__wrapper">
          <Row isNoGutters={true} className="flex-nowrap" role="presentation">
            <Col isFill={true} role="presentation">
              <div className="search__input" role="search">
                <Input
                  value={value}
                  onChange={this.handleOnChange}
                  onEnterPress={this.handleOnSubmit}
                  onKeyPress={onKeyPress}
                  onFocus={onFocus}
                  onBlur={onBlur}
                  placeholder={placeholder}
                  autoFocus={isVisible}
                  ariaLabel={ariaLabel}
                  htmlType="search"
                  name="search"
                />
                <SearchClearButton value={value} onClear={this.handleOnClear} />
                <SearchSubmitButton
                  value={value}
                  onSubmit={this.handleOnSubmit}
                  isSubmitEnabledAlways={isSubmitEnabledAlways}
                />
              </div>
            </Col>
            {onClose && (
              <Col isAuto={true} role="presentation">
                <SearchCloseButton onClick={this.handleOnClose} />
              </Col>
            )}
          </Row>
        </div>
      </div>
    );
  }
}

export default Search;
