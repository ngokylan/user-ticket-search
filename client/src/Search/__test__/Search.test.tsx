import React from 'react';
import { mount } from 'enzyme';
import Search, { SearchProps } from '..';

const defaultProps: SearchProps = {
  placeholder: 'Search',
  ariaLabel: 'Search field',
};
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Search {...props} />);
  const component = wrapper.find('.search');

  return { props, wrapper, component };
};

describe('Search', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it.each([
      ['Heading exist', { heading: 'Title' }],
      ['Description exist', { description: 'Description' }],
    ])('should be match with snapshot %s', (name, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Search.defaultProps)
        );
      });

      it.each([
        ['search-default', {}],
        ['search-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });

    describe('custom', () => {
      it.each([['id', 'test', { id: 'test' }]])(
        'should render prop %s as %s',
        (name, expected, props) => {
          const { component } = mountComponent(props);
          expect(component.prop(name)).toBe(expected);
        }
      );

      it.each([
        ['value', 'some-value', { value: 'some-value' }],
        ['value', '', { value: '' }],
        ['value', '', {}],
        ['onFocus', undefined, {}],
        ['onFocus', true, { onFocus: true }],
        ['onFocus', false, { onFocus: false }],
        ['onBlur', undefined, {}],
        ['onBlur', true, { onBlur: true }],
        ['onBlur', false, { onBlur: false }],
        ['placeholder', defaultProps.placeholder, {}],
        [
          'placeholder',
          'some-placeholder',
          { placeholder: 'some-placeholder' },
        ],
        ['autoFocus', undefined, {}],
        ['autoFocus', true, { isVisible: true }],
        ['autoFocus', false, { isVisible: false }],
      ])('should render prop %s as %s for Input', (name, expected, props) => {
        const { component } = mountComponent(props);
        expect(component.find('Input').prop(name)).toBe(expected);
      });

      it.each([
        ['value', 'some-value', { value: 'some-value' }],
        ['value', '', { value: '' }],
        ['value', '', {}],
      ])('should render prop %s as %s for Input', (name, expected, props) => {
        const { component } = mountComponent(props);
        expect(component.find('SearchClearButton').prop(name)).toBe(expected);
      });

      it.each([
        ['value', 'some-value', { value: 'some-value' }],
        ['value', '', { value: '' }],
        ['value', '', {}],
      ])('should render prop %s as %s for Input', (name, expected, props) => {
        const { component } = mountComponent(props);
        expect(component.find('SearchSubmitButton').prop(name)).toBe(expected);
      });
    });
  });

  describe('interaction', () => {
    it.each([
      ['should', '.search', 1, {}],
      ['should', '.search--open', 1, { isVisible: true }],
      ['should not', '.search--open', 0, { isVisible: false }],
      ['should not', 'SearchCloseButton', 0, {}],
      ['should', 'SearchCloseButton', 1, { onClose: true }],
      ['should not', 'SearchCloseButton', 0, { onClose: false }],
    ])('%s have %s', (name, element, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper.find(element)).toHaveLength(expected);
    });
  });

  describe('api', () => {
    it('should have call handleOnClose once', () => {
      jest.clearAllMocks();
      const props = {
        onClose: jest.fn(),
      };
      const search = new Search(props);
      search.handleOnClose();
      expect(props.onClose.mock.calls.length).toEqual(1);
    });

    it('should have call handleOnSubmit once', () => {
      jest.clearAllMocks();
      const props = {
        onSubmit: jest.fn(),
      };
      const search = new Search(props);
      search.handleOnSubmit();
      expect(props.onSubmit.mock.calls.length).toEqual(1);
    });

    it('should have call handleOnChange once', () => {
      jest.clearAllMocks();
      const props = {};
      const search = new Search(props);
      const event = { target: { value: 'some-value' } };
      search.setValue = jest.fn();

      search.handleOnChange(event);
      expect(search.setValue.mock.calls.length).toEqual(1);
      expect(search.setValue).toBeCalledWith(event.target.value);
    });

    it('should have call handleOnClear once', () => {
      jest.clearAllMocks();
      const props = {};
      const search = new Search(props);
      search.setValue = jest.fn();

      search.handleOnClear();
      expect(search.setValue.mock.calls.length).toEqual(1);
      expect(search.setValue).toBeCalledWith('');
    });

    it('should have call setValue callback function properly', () => {
      jest.clearAllMocks();
      const props = {
        onChange: jest.fn(),
      };
      const search = new Search(props);

      search.setValue('some-value');
      expect(props.onChange.mock.calls.length).toEqual(1);
      expect(props.onChange).toBeCalledWith('some-value');

      search.setValue('another value');
      expect(props.onChange.mock.calls.length).toEqual(2);
      expect(props.onChange).toBeCalledWith('another value');
    });
  });
});
