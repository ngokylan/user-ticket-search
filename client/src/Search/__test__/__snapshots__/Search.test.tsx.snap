// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search rendering should be match with snapshot Description exist 1`] = `
<div
  className="search"
  data-testid="search-default"
>
  <div
    className="search__wrapper"
  >
    <div
      className="row flex-nowrap no-gutters"
      role="presentation"
    >
      <div
        className="col flex-fill"
        role="presentation"
      >
        <div
          className="search__input"
          role="search"
        >
          <input
            aria-invalid={false}
            aria-label="Search field"
            className="input"
            data-testid="input-default"
            disabled={false}
            hidden={false}
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            onMouseDown={[Function]}
            placeholder="Search"
            readOnly={false}
            tabIndex={0}
            type="search"
            value=""
          />
          <button
            className="sc-bdfBwQ gjsWpT btn search__btn-submit btn--default btn--icon btn--is-uppercase btn--is-text btn--is-disabled"
            data-testid="btn-default"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            role="button"
            type="button"
          >
            <span
              className="btn__icon"
              data-testid="btn-icon-default"
            >
              <span
                className="icon icon--size-md"
                data-testid="icon"
                title="Search"
              >
                <svg
                  aria-hidden="true"
                  className="icon__svg"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <g>
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                  </g>
                </svg>
              </span>
            </span>
            <span
              className="btn__hover"
            />
            <span
              className="btn__focused"
            />
            <span
              className="btn__pressed"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search rendering should be match with snapshot Heading exist 1`] = `
<div
  className="search"
  data-testid="search-default"
>
  <div
    className="search__wrapper"
  >
    <div
      className="row flex-nowrap no-gutters"
      role="presentation"
    >
      <div
        className="col flex-fill"
        role="presentation"
      >
        <div
          className="search__input"
          role="search"
        >
          <input
            aria-invalid={false}
            aria-label="Search field"
            className="input"
            data-testid="input-default"
            disabled={false}
            hidden={false}
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            onMouseDown={[Function]}
            placeholder="Search"
            readOnly={false}
            tabIndex={0}
            type="search"
            value=""
          />
          <button
            className="sc-bdfBwQ gjsWpT btn search__btn-submit btn--default btn--icon btn--is-uppercase btn--is-text btn--is-disabled"
            data-testid="btn-default"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            role="button"
            type="button"
          >
            <span
              className="btn__icon"
              data-testid="btn-icon-default"
            >
              <span
                className="icon icon--size-md"
                data-testid="icon"
                title="Search"
              >
                <svg
                  aria-hidden="true"
                  className="icon__svg"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <g>
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                  </g>
                </svg>
              </span>
            </span>
            <span
              className="btn__hover"
            />
            <span
              className="btn__focused"
            />
            <span
              className="btn__pressed"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
